[BEAM-8910] Make custom BQ source read from Avro

pabloem opened the pull request:
This adds support for reading data from BigQuery using FastAvro in the new BigQuery source transform: `_ReadFromBigQuery`.

This creates a few subtle changes in the way that data is returned by BigQuery. These changes are the following:

- All the time-specific data types in BigQuery (`DATE`, `DATETIME`, `TIMESTAMP`) are converted into Python-specific types (`datetime.date`, `datetime.datetime`, `datetime.datetime`).
  - This is different from `BigQuerySource`, where we export JSON data, and all of these datatypes are returned as strings.

- The BigQuery `BYTES` type is returned as Python `bytes`.
  - This is different from `BigQuerySource`, where we export JSON data. For JSON, BigQuery (and beam's BigQuerySource) returns base64-encoded bytes.

Note that in tests, the behavior of BigQuerySource **remains the same**, while _ReadFromBigQuery has different behavior.

This transform does not need to honor backwards compatibility, as it is a new transform, and **it will not replace** BigQuerySource automatically. Nonetheless, we should definitely note the subtle API differences for it.

r: @chamikaramj @tvalentyn @kamilwu 

cc: @mwalenia 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python2_PVR_Flink PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
@chamikaramj @tvalentyn 

Hey there, I'm working on this improvement. There's some interesting outcomes here:

- For datetime types (date,time,datetime,timestamp), AvroSource returns python types (datetime.date, datetime.time,datetime.datetime, datetime.datetime) - along with timezone info. I feel that in this case, Python types are the most desirable. Thoughts? _CustomBQSource is experimental, so I wouldn't think it's subject to backwards compatibility concerns.

- For bytes, AvroSource is currently returning base64 encoded bytes. I am trying to figure out how to handle this. For py2, strings and bytes may not be differentiable, so we may not be able to handle this properly...

tvalentyn commented:
> * For bytes, AvroSource is currently returning base64 encoded bytes. I am trying to figure out how to handle this. For py2, strings and bytes may not be differentiable, so we may not be able to handle this properly...
Do you have a summary of Py2 vs Py3  x Current Behavior vs New behavior? What is returned on Py3 - is it also base64 encoded bytes?
This is a prior discussion on Py3 migration issues in Bigquery: https://lists.apache.org/thread.html/f35c836887014e059527ed1a806e730321e2f9726164a3030575f455%40%3Cdev.beam.apache.org%3E. It took us some time to understand what was going on (https://lists.apache.org/thread.html/0c2178cf8e5d9e77c4f233f05a0b87b6011a1daa1a5ae47b41463af5%40%3Cdev.beam.apache.org%3E)

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
@kamilwu @chamikaramj PTAL : ) -- see first comment for analysis on typing, and changes

tvalentyn commented on sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py:
```diff
@@ -236,11 +251,12 @@ def create_table(cls, table_name):
         'float': 0.33,
         'numeric': Decimal('10'),
-        'bytes': base64.b64encode(b'\xab\xac'),
+        'bytes': base64.b64encode(byts) if native else byts,
```
How intuitive will this behavior be to users? should't the output be consistent? 
If we decide to keep it, how will this be documented?

pabloem replied:
The behavior will be different for different transforms. Users will need to explicitly change the transform in their code. We can make them aware of the new transform, and its typing differences in release notes, and possibly in Pydoc for the new transform as well. Thoughts?

tvalentyn replied:
Bytes treatment should be called out in IO doc, we do mention it: https://github.com/apache/beam/blob/8bc2880cca40c00a96623b3ce96ea0b856af76c9/sdks/python/apache_beam/io/gcp/bigquery.py#L227. 

b64encoding may be unnecessary, and less efficient. I think the reason native IO encodes bytes with b64 is because that was the behavior in Java SDK. We can argue that's not necessary. However I am concerned about the consistency of the UX here. Different UX for two transforms means the transforms will not be interchangeable, and users might overlook this.  This might also cause friction in cross-language pipelines.

pabloem replied:
I think that's reasonable. But I'd think that after slight confusion, users would figure out the differences.

Another option is to provide a parameter to the transform to have backwards-compatible types at a performance cost.

@chamikaramj @tvalentyn  wdyt?

tvalentyn replied:
Having a parameter for backwards-compatibility would make sense to me, but will defer to @chamikaramj  on this. Since new transform comes with a new name, whoever will be using it, will have to read the API, so will be aware of the issue. But that would mean we add a knob, potentially indefinitely.

pabloem replied:
So the proposal right now is: To have a parameter for backwards compatibility in the transform, but by default return Avro-returned types. We will need to have clear Pydoc on _ReadFromBigQuery, and on CHANGES.md.
@chamikaramj thoughts?

pabloem replied:
@chamikaramj @tvalentyn @kamilwu - added a backward compatibility flag. Can you review?

pabloem replied:
If the changes are reasonable, I'll add to release notes

chamikaramj replied:
I think having a single "backwards compatibility" parameter can be confusing and error prone. We might still be backwards incompatible in some unknown ways. I think it's better to document exact differences between the transforms and give clear documentation/guidelines for anyone who is upgrading.

kamilwu commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -663,14 +662,10 @@ def split(self, desired_bundle_size, start_position=None, stop_position=None):
         self.table_reference = self._execute_query(bq)
 
-      schema, metadata_list = self._export_files(bq)
+      unused_schema, metadata_list = self._export_files(bq)
```
Does it mean that coder is no longer needed for `_CustomBigQuerySource`?

pabloem replied:
You're correct!

kamilwu replied:
This is great! That would mean that we can safely remove `_JsonToDictCoder`. Do you think you can do this as a part of this PR?

pabloem replied:
We may do that, but if we end up keeping a backwards compatibility flag, we'll need to keep the coder as-is.

kamilwu replied:
I see. Then let's wait for the decision.

kamilwu commented on sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py:
```diff
@@ -69,6 +70,20 @@ def wrapped(self):
 
+def datetime_to_utc(element):
+  for k, v in element.items():
+    if isinstance(v, datetime.time):
```
We can combine this condition with the one below:
`if isinstance(v, datetime.time) or isinstance(v, datetime.date):`

robertwb replied:
or `isinstance(v, (datetime.time, datetime.date))`

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Py35 postcommit run: https://builds.apache.org/job/beam_PostCommit_Python35_PR/114/

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Passing py35 pc: https://builds.apache.org/job/beam_PostCommit_Python35_PR/118/

pabloem commented:
Lint failures unrelated

pabloem commented:
Precommit failures unrelated.

tvalentyn commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -609,7 +611,8 @@ def __init__(
       flatten_results=True,
-      kms_key=None):
+      kms_key=None,
+      backwards_compatible_data_format=None):
```
1. Can we use a bool as a default value instead of `None`?
2. Can we use rename this flag to something that describes the course of action, for example: `use_json_data_representation=False` - you can come up with a better name.

pabloem replied:
1. Done.
2. I've called it `use_json_exports` - does that help?

tvalentyn commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -1570,6 +1584,10 @@ class _ReadFromBigQuery(PTransform):
+    backwards_compatible_data_format (bool): By default, this transform reads
+      data in native Python types that come from Avro-exports of BigQuery. By
+      setting this flag to True, the transform will return JSON-types, like
+      the older BigQuerySource.
```
1. Do all BQ types have a native python type?
2. What are JSON-types?

pabloem replied:
I've added a link to https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-avro#avro_conversions for details on the conversions. I've also rephrased the doc. LMK what you think.

pabloem commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -610,7 +611,8 @@ def __init__(
       flatten_results=True,
-      kms_key=None):
+      kms_key=None,
+      use_json_exports=False):
```
not supporting JSON would increase the amount of changes that users need to make to migrate to the new source, as the data output types would change. I don't think this would be ideal. Wouldn't it be better to allow users to migrate to the new source without having to change the rest of their pipeline?

robertwb replied:
Assuming Avro is a superior format, I think we should drop support for json (but not right away). My preference would be to have this choice be orthogonal to the choice of types (and having the bytes default to be base64 encoded feels really wrong) so do think having a backwards compatibility parameter is preferable. (If it's just the date and bytes types, I would split this into two parameters: `use_base64_encoded_bytes` and `datedtime_type`.)

tvalentyn replied:
> "having the bytes default to be base64 encoded feels really wrong" .
I think we changed bytes to be base64-encoded several years back, because Java SDK used that representation: https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L181-L182

lukecwik replied:
Yes, the joys of matching what was done in Java years ago for the Dataflow service API.

pabloem replied:
Restarting this discussion after some rebasing and preparing...

I think I'd prefer to allow users to choose the export file format. This is what we do for writing to BQ in Java and Python. Allowing users to choose output type for specific column types would add overhead of (if avro and json_format_bytes: formatbytes; if avro and json_datetime_format: formatdatetime; if not avro and not json_format_bytes: formatbytes; ...).

We can discourage users from using JSON (AVRO is already the default) - and eventually stop supporting it if that makes sense. Thoughts?

pabloem commented:
@chamikaramj @tvalentyn @kamilwu 
I've documented the data types that are different, and provided links to BQ documentation to see differences in representation. I've also addressed Cham's comments. I think it's worth keeping JSON export capability.

Finally, I've added the change to CHANGES.md. This should help us move the source forward towards full support. Thoughts?

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python2_PVR_Flink PreCommit

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
@chamikaramj @tvalentyn @kamilwu PTAL - see comment https://github.com/apache/beam/pull/11086#issuecomment-615517971

tvalentyn commented:
Thanks Pablo.
1. Can we take another look at a big comment at the beginning of bigquery.py and see if it needs an update. It sounds like we now have two ways of reading and writing to BigQuery. Can we add a guidance to users when to use which, and what they will need to change when they switch from one to another?

2. The same comment says: BigQuery IO requires values of BYTES datatype to be encoded using base64
encoding when writing to BigQuery. When bytes are read from BigQuery they are
returned as base64-encoded bytes. Does this apply to the new source? Sounds like not, so please clarify.
3. What is the test story for BQ IO? I see several branches here: natve vs non-native IO, Avro vs Json. Which combinations are tested?
4. Big Query tests are one of the major source of [postcommit flakiness](https://issues.apache.org/jira/issues/?jql=text~BigQuery%20AND%20summary~flaky%20AND%20status%3DOpen%20AND%20summary~Test)? What are the plans to address that?

pabloem commented:
> Thanks Pablo.
> 
> 1. Can we take another look at a big comment at the beginning of bigquery.py and see if it needs an update. It sounds like we now have two ways of reading and writing to BigQuery. Can we add a guidance to users when to use which, and what they will need to change when they switch from one to another?
> 2. The same comment says: BigQuery IO requires values of BYTES datatype to be encoded using base64
>    encoding when writing to BigQuery. When bytes are read from BigQuery they are
>    returned as base64-encoded bytes. Does this apply to the new source? Sounds like not, so please clarify.

Ah great observation. Thanks Valentyn. I've rephrased that. LMK what you think now

> 3. What is the test story for BQ IO? I see several branches here: natve vs non-native IO, Avro vs Json. Which combinations are tested?

There are tests for pretty much all the variants. There are:
- Native: `BigQuerySource`. This is widely tested and used.
- Custom (AVRO) `ReadFromBigQuery`. This is tested by almost all of the BQ-related PostCommits.
- Custom (JSON) `ReadFromBigQuery(use_json_exports=True)`. This is tested in a couple of the BQ-related PostCommits (note `BigQueryQueryToTableIT.test_new_types_json` checks all the special types including bytes, and datetime-related types).

> 4. Big Query tests are one of the major source of [postcommit flakiness](https://issues.apache.org/jira/issues/?jql=text~BigQuery%20AND%20summary~flaky%20AND%20status%3DOpen%20AND%20summary~Test)? What are the plans to address that?

These are all Write tests. The Read from BQ tests are less troublesome, likely due to the fact that the transform is less complex, and asserts are kept in the pipeline (vs BigQuery matchers). I'm happy to own all breakages related to thistest - though I expect few compared to the WriteToBQ issues.

pabloem commented:
Run Python 3.7 PostCommit

tvalentyn commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -45,8 +45,8 @@
 
-  main_table = pipeline | 'VeryBig' >> beam.io.Read(beam.io.BigQuerySource()
-  side_table = pipeline | 'NotBig' >> beam.io.Read(beam.io.BigQuerySource()
+  main_table = pipeline | 'VeryBig' >> beam.io.ReadFroBigQuery(...)
```
Typo in From

tvalentyn replied:
Ping on this

tvalentyn commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -78,6 +72,12 @@
 or a table. Pipeline construction will fail with a validation error if neither
 or both are specified.
 
+When reading from BigQuery using `BigQuerySource`, bytes are returned as
```
Can we add some guidance when to use which?

tvalentyn replied:
Ping on this - do we need to guide users in any way which transform they should use? Is some of them experimental or to be deprecated?

tvalentyn commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run PythonDocker PreCommit

pabloem commented:
Run Python2_PVR_Flink PreCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.6 PostCommit

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Website_Stage_GCS PreCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python2_PVR_Flink PreCommit

tvalentyn commented:
@pabloem, would you mind squashing the commits that happened after the last
review iteration (but please don't squash earlier commits until review is
finalized) and resolving any comments that are already addressed? You can
also delete the "Run postcommit..." which would make  it easier to follow
up on remaining comment threads.

On Thu, May 21, 2020 at 12:52 PM Pablo <notifications@github.com> wrote:

> *@pabloem* commented on this pull request.
> ------------------------------
>
> In sdks/python/apache_beam/io/gcp/bigquery.py
> <https://github.com/apache/beam/pull/11086#discussion_r428877884>:
>
> > @@ -610,7 +611,8 @@ def __init__(
>        coder=None,
>        use_standard_sql=False,
>        flatten_results=True,
> -      kms_key=None):
> +      kms_key=None,
> +      use_json_exports=False):
>
> Restarting this discussion after some rebasing and preparing...
>
> I think I'd prefer to allow users to choose the export file format. This
> is what we do for writing to BQ in Java and Python. Allowing users to
> choose output type for specific column types would add overhead of (if avro
> and json_format_bytes: formatbytes; if avro and json_datetime_format:
> formatdatetime; if not avro and not json_format_bytes: formatbytes; ...).
>
> We can discourage users from using JSON (AVRO is already the default) -
> and eventually stop supporting it if that makes sense. Thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/11086#discussion_r428877884>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHZCYCRXV5WRILNRQP2IZDRSWBA3ANCNFSM4LESJLFQ>
> .
>

pabloem commented:
Run Python 3.5 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
py2 failure is hdfs integration test

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python2_PVR_Flink PreCommit

pabloem commented on sdks/python/apache_beam/io/gcp/big_query_query_to_table_it_test.py:
```diff
@@ -254,11 +256,36 @@ def test_big_query_new_types(self):
     big_query_query_to_table_pipeline.run_bq_pipeline(options)
 
+  @attr('IT')
+  def test_big_query_new_types(self):
```
Thanks for pointing that out. I've renamed the test cases to keep coverage for the feature on the same test.

pabloem commented:
So, I think the last remaining issue was whether to have the `use_json_exports` flag or individual flags per columns. Are there other topics to discuss?

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.7 PostCommit

pabloem commented:
@tvalentyn @chamikaramj thoughts?

tvalentyn commented:
I will defer to Cham on use_json_exports.

chamikaramj commented:
Can you provide some more context :) ?

chamikaramj commented:
Based on some offline discussion, I think it's enough to add a single flag here. We should try to minimize the number of flags we add (add flags is easy removing is hard).

The main question here I think is regarding preserving backwards compatibility for users who rely on old behavior. For this purpose a single flag is enough. Users who need to convert a given type from one form to the other can easily do that by adding a ParDo. If adding support for a certain type is beneficial to a lot of users and if there's significant user demand we can always add that later.

pabloem commented:
okay, in that case, the main points have been addressed. Any objections from merging? fyi @chamikaramj @tvalentyn @robertwb @kamilwu 

If no objections, I'll merge in a day or two.

tvalentyn commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -732,11 +748,20 @@ def _export_files(self, bq):
+                                       bigquery_tools.FileFormat.JSON,
+                                       include_header=False)
+    else:
+      job_ref = bq.perform_extract_job([self.gcs_location],
```
Is this resolved?

tvalentyn commented on sdks/python/apache_beam/io/gcp/bigquery.py:
```diff
@@ -78,6 +72,12 @@
+When reading from BigQuery using `BigQuerySource`, bytes are returned as
+base64-encoded bytes. When reading via `ReadFromBigQuery`, bytes are returned
+as bytes without base64 encoding. This is due to the fact that ReadFromBigQuery
+uses Avro expors by default. To get base64-encoded bytes, you can use the flag
```
nit: exports

pabloem replied:
done

tvalentyn commented on CHANGES.md:
```diff
@@ -58,6 +58,10 @@
 
 * Support for X source added (Java/Python) ([BEAM-X](https://issues.apache.org/jira/browse/BEAM-X)).
 * Support for reading from Snowflake added (Java) ([BEAM-9722](https://issues.apache.org/jira/browse/BEAM-9722)).
+* A new transform to read from BigQuery has been added: `apache_beam.io.gcp.bigquery.ReadFromBigQuery`. This transform
```
To confirm - there is no changes in behavior for old sync, right?

pabloem replied:
that's correct.

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Portable_Python PreCommit

pabloem commented:
Run Python2_PVR_Flink PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Run Python2_PVR_Flink PreCommit

pabloem commented:
Run Python 2 PostCommit

pabloem commented:
Run Python 3.6 PostCommit
